File: historian/src/query/engine.ts
import fs from "node:fs/promises";
import path from "node:path";
import { HistorianConfig } from "../config/types";
import { indexPath, segmentPath, shardForTag } from "../storage/layout";
import { readBlockIndex } from "../storage/readers";
import { decodeSegmentRecord } from "../storage/codec";
import { isNumericType } from "../types/valueTypes";
import { AggName, LastQueryRequest, QueryOrder, QueryPoint, RangeQueryRequest, RawQueryRequest } from "./types";

interface ScanRange {
  start: bigint;
  end: bigint;
}

interface BucketAggState {
  first: QueryPoint | null;
  last: QueryPoint | null;
  min: number | null;
  max: number | null;
  sum: number;
  numericCount: number;
  count: number;
  sawString: boolean;
}

function partitionStarts(config: HistorianConfig, from: bigint, to: bigint): number[] {
  const fromMs = config.storage.timestampUnit === "ns" ? Number(from / 1_000_000n) : Number(from / 1_000n);
  const toMs = config.storage.timestampUnit === "ns" ? Number(to / 1_000_000n) : Number(to / 1_000n);
  const step = config.storage.partitionDurationMs;
  const start = Math.floor(fromMs / step) * step;
  const out: number[] = [];
  for (let ms = start; ms <= toMs; ms += step) out.push(ms);
  return out;
}

async function blockScanRanges(
  config: HistorianConfig,
  dataDir: string,
  partitionStartMs: number,
  shard: number,
  tagMin: number,
  tagMax: number,
  from: bigint,
  to: bigint
): Promise<{ segPath: string; ranges: ScanRange[] }> {
  const dt = new Date(partitionStartMs);
  const day = `${dt.getUTCFullYear()}-${(dt.getUTCMonth() + 1).toString().padStart(2, "0")}-${dt.getUTCDate().toString().padStart(2, "0")}`;
  const hour = dt.getUTCHours().toString().padStart(2, "0");
  const partition = { day, hour, folder: `${day}/${hour}`, partitionStartMs };
  const segPath = segmentPath(dataDir, partition, shard);
  const idxPath = indexPath(dataDir, partition, shard);

  const idxEntries = await readBlockIndex(idxPath);
  if (idxEntries.length === 0) {
    try {
      const st = await fs.stat(segPath);
      if (st.size === 0) return { segPath, ranges: [] };
      return { segPath, ranges: [{ start: 0n, end: BigInt(st.size) }] };
    } catch {
      return { segPath, ranges: [] };
    }
  }

  const ranges: ScanRange[] = [];
  for (const e of idxEntries) {
    if (e.maxTs < from || e.minTs > to) continue;
    if (e.maxTagId < tagMin || e.minTagId > tagMax) continue;
    ranges.push({ start: e.byteOffsetStart, end: e.byteOffsetEnd });
  }
  return { segPath, ranges };
}

function createBucketState(): BucketAggState {
  return {
    first: null,
    last: null,
    min: null,
    max: null,
    sum: 0,
    numericCount: 0,
    count: 0,
    sawString: false
  };
}

function updateBucket(state: BucketAggState, point: QueryPoint): void {
  state.count += 1;
  if (!state.first || point.tsEpoch < state.first.tsEpoch) state.first = point;
  if (!state.last || point.tsEpoch > state.last.tsEpoch) state.last = point;
  if (!isNumericType(point.typeCode)) {
    state.sawString = true;
    return;
  }
  const value = point.value as number;
  state.numericCount += 1;
  state.sum += value;
  state.min = state.min === null ? value : Math.min(state.min, value);
  state.max = state.max === null ? value : Math.max(state.max, value);
}

function finalizeBucket(state: BucketAggState, agg: AggName): number | string | null {
  switch (agg) {
    case "count":
      return state.count;
    case "first":
      return state.first ? String(state.first.value) : null;
    case "last":
      return state.last ? String(state.last.value) : null;
    case "min":
      return state.sawString ? null : state.min;
    case "max":
      return state.sawString ? null : state.max;
    case "avg":
      return state.sawString || state.numericCount === 0 ? null : state.sum / state.numericCount;
    default:
      return null;
  }
}

export class QueryEngine {
  constructor(private readonly config: HistorianConfig) {}

  async last(req: LastQueryRequest): Promise<{ points: QueryPoint[] }> {
    const tagSet = new Set(req.tagIds);
    const sortedTags = [...tagSet].sort((a, b) => a - b);
    const tagMin = sortedTags[0];
    const tagMax = sortedTags[sortedTags.length - 1];
    const byShard = new Map<number, number[]>();
    for (const t of tagSet) {
      const shard = shardForTag(t, this.config.storage.shardCount);
      const arr = byShard.get(shard) ?? [];
      arr.push(t);
      byShard.set(shard, arr);
    }

    const lastByTag = new Map<number, QueryPoint>();
    const partitions = await this.listPartitionsDescending();
    for (const pStart of partitions) {
      for (const shard of byShard.keys()) {
        const scan = await blockScanRanges(
          this.config,
          this.config.storage.dataDir,
          pStart,
          shard,
          tagMin,
          tagMax,
          -9223372036854775808n,
          9223372036854775807n
        );
        if (scan.ranges.length === 0) continue;
        for (let i = scan.ranges.length - 1; i >= 0; i--) {
          const range = scan.ranges[i];
          const chunk = await this.readRange(scan.segPath, range.start, range.end);
          let off = 0;
          while (off < chunk.length) {
            const rec = decodeSegmentRecord(chunk, off);
            if (!rec) break;
            off += rec.bytesRead;
            const p = rec.point;
            if (!tagSet.has(p.tagId)) continue;
            const cur = lastByTag.get(p.tagId);
            if (!cur || p.tsEpoch > cur.tsEpoch) lastByTag.set(p.tagId, p);
          }
          if (lastByTag.size >= tagSet.size) break;
        }
        if (lastByTag.size >= tagSet.size) break;
      }
      if (lastByTag.size >= tagSet.size) break;
    }
    return { points: [...lastByTag.values()] };
  }

  async raw(req: RawQueryRequest): Promise<{ points: QueryPoint[]; truncated: boolean }> {
    const order: QueryOrder = req.order ?? "desc";
    const tagSet = new Set(req.tagIds);
    const sortedTags = [...tagSet].sort((a, b) => a - b);
    const tagMin = sortedTags[0];
    const tagMax = sortedTags[sortedTags.length - 1];
    const pStarts = partitionStarts(this.config, req.from, req.to);
    if (order === "desc") pStarts.reverse();
    const byShard = new Map<number, number[]>();
    for (const t of tagSet) {
      const shard = shardForTag(t, this.config.storage.shardCount);
      const arr = byShard.get(shard) ?? [];
      arr.push(t);
      byShard.set(shard, arr);
    }
    const out: QueryPoint[] = [];
    for (const pStart of pStarts) {
      for (const shard of byShard.keys()) {
        const scan = await blockScanRanges(
          this.config,
          this.config.storage.dataDir,
          pStart,
          shard,
          tagMin,
          tagMax,
          req.from,
          req.to
        );
        if (scan.ranges.length === 0) continue;
        const left = req.limit - out.length;
        if (left <= 0) return { points: out, truncated: true };
        const subset = await this.collectFromRanges(scan.segPath, scan.ranges, tagSet, req.from, req.to, left, order);
        out.push(...subset);
        if (out.length >= req.limit) return { points: out, truncated: true };
      }
    }
    out.sort((a, b) => {
      if (a.tsEpoch === b.tsEpoch) return a.tagId - b.tagId;
      return order === "desc" ? (a.tsEpoch > b.tsEpoch ? -1 : 1) : a.tsEpoch < b.tsEpoch ? -1 : 1;
    });
    return { points: out, truncated: false };
  }

  async range(req: RangeQueryRequest): Promise<{ buckets?: unknown; points?: QueryPoint[]; truncated?: boolean }> {
    if (!req.bucketMs) {
      return this.raw(req);
    }
    const raw = await this.raw(req);
    const bucketUs = BigInt(req.bucketMs) * (this.config.storage.timestampUnit === "ns" ? 1_000_000n : 1_000n);
    const state = new Map<number, Map<string, BucketAggState>>();
    for (const p of raw.points) {
      const d = state.get(p.tagId) ?? new Map<string, BucketAggState>();
      state.set(p.tagId, d);
      const key = ((p.tsEpoch - req.from) / bucketUs).toString();
      const s = d.get(key) ?? createBucketState();
      updateBucket(s, p);
      d.set(key, s);
    }
    const result: Record<string, Array<{ bucketStart: string; value: number | string | null }>> = {};
    for (const [tagId, buckets] of state) {
      const rows: Array<{ bucketStart: string; value: number | string | null }> = [];
      for (const [k, s] of buckets) {
        const start = req.from + BigInt(k) * bucketUs;
        rows.push({ bucketStart: start.toString(), value: finalizeBucket(s, req.agg) });
      }
      rows.sort((a, b) => {
        if (a.bucketStart === b.bucketStart) return 0;
        if ((req.order ?? "desc") === "desc") return BigInt(a.bucketStart) > BigInt(b.bucketStart) ? -1 : 1;
        return BigInt(a.bucketStart) < BigInt(b.bucketStart) ? -1 : 1;
      });
      result[String(tagId)] = rows;
    }
    return { buckets: result, truncated: raw.truncated };
  }

  private async listPartitionsDescending(): Promise<number[]> {
    const rawDir = path.join(this.config.storage.dataDir, "raw");
    const out: number[] = [];
    let days: Array<{ name: string; isDirectory: () => boolean }> = [];
    try {
      days = await fs.readdir(rawDir, { withFileTypes: true });
    } catch {
      return out;
    }
    for (const day of days) {
      if (!day.isDirectory()) continue;
      const dayPath = path.join(rawDir, day.name);
      let hours: Array<{ name: string; isDirectory: () => boolean }> = [];
      try {
        hours = await fs.readdir(dayPath, { withFileTypes: true });
      } catch {
        continue;
      }
      for (const hour of hours) {
        if (!hour.isDirectory()) continue;
        const dt = new Date(`${day.name}T${hour.name}:00:00.000Z`);
        if (!Number.isFinite(dt.getTime())) continue;
        out.push(dt.getTime());
      }
    }
    out.sort((a, b) => b - a);
    return out;
  }

  private async readRange(segPath: string, start: bigint, end: bigint): Promise<Buffer> {
    const len = Number(end - start);
    if (len <= 0) return Buffer.alloc(0);
    const fh = await fs.open(segPath, "r");
    try {
      const buf = Buffer.allocUnsafe(len);
      const rd = await fh.read(buf, 0, len, Number(start));
      return rd.buffer.subarray(0, rd.bytesRead);
    } finally {
      await fh.close();
    }
  }

  private async collectFromRanges(
    segPath: string,
    ranges: ScanRange[],
    tagSet: Set<number>,
    from: bigint,
    to: bigint,
    limit: number,
    order: QueryOrder
  ): Promise<QueryPoint[]> {
    const out: QueryPoint[] = [];
    const fh = await fs.open(segPath, "r");
    try {
      const rangesIter = order === "desc" ? [...ranges].reverse() : ranges;
      for (const r of rangesIter) {
        const len = Number(r.end - r.start);
        if (len <= 0) continue;
        const buf = Buffer.allocUnsafe(len);
        const rd = await fh.read(buf, 0, len, Number(r.start));
        const chunk = rd.buffer.subarray(0, rd.bytesRead);
        if (order === "asc") {
          let off = 0;
          while (off < chunk.length) {
            const rec = decodeSegmentRecord(chunk, off);
            if (!rec) break;
            off += rec.bytesRead;
            const p = rec.point;
            if (!tagSet.has(p.tagId)) continue;
            if (p.tsEpoch < from || p.tsEpoch > to) continue;
            out.push(p);
            if (out.length >= limit) return out;
          }
        } else {
          const pointsInBlock: QueryPoint[] = [];
          let off = 0;
          while (off < chunk.length) {
            const rec = decodeSegmentRecord(chunk, off);
            if (!rec) break;
            off += rec.bytesRead;
            const p = rec.point;
            if (!tagSet.has(p.tagId)) continue;
            if (p.tsEpoch < from || p.tsEpoch > to) continue;
            pointsInBlock.push(p);
          }
          for (let i = pointsInBlock.length - 1; i >= 0; i--) {
            out.push(pointsInBlock[i]);
            if (out.length >= limit) return out;
          }
        }
      }
      return out;
    } finally {
      await fh.close();
    }
  }
}

File: historian_v2/src/query/engine.rs
use anyhow::Result;
use serde::Serialize;
use std::collections::{BTreeMap, HashMap, HashSet};
use std::path::Path;

use crate::config::{Config, TimestampUnit};
use crate::storage::codec::{decode_block_index, decode_segment_record};
use crate::storage::layout::{shard_for_tag, ts_to_ms, PartitionInfo};
use crate::types::{Point, PointValue, QueryOrder};

#[derive(Debug, Clone, Copy)]
pub enum AggName {
    Min,
    Max,
    Avg,
    First,
    Last,
    Count,
}

impl AggName {
    pub fn from_str(v: Option<&str>) -> Result<Self> {
        Ok(match v.unwrap_or("avg") {
            "min" => Self::Min,
            "max" => Self::Max,
            "avg" => Self::Avg,
            "first" => Self::First,
            "last" => Self::Last,
            "count" => Self::Count,
            _ => anyhow::bail!("agg is invalid"),
        })
    }
}

#[derive(Clone)]
pub struct QueryEngine {
    cfg: Config,
}

#[derive(Debug, Serialize)]
pub struct RawResult {
    pub points: Vec<Point>,
    pub truncated: bool,
}

#[derive(Debug, Serialize)]
pub struct RangeResult {
    pub buckets: HashMap<String, Vec<BucketValue>>,
    pub truncated: bool,
}

#[derive(Debug, Serialize)]
pub struct BucketValue {
    #[serde(rename = "bucketStart")]
    pub bucket_start: String,
    pub value: serde_json::Value,
}

impl QueryEngine {
    pub fn new(cfg: Config) -> Self {
        Self { cfg }
    }

    pub async fn raw(
        &self,
        tag_ids: &[u32],
        from: i64,
        to: i64,
        limit: usize,
        order: QueryOrder,
    ) -> Result<RawResult> {
        let tag_set: HashSet<u32> = tag_ids.iter().copied().collect();
        let mut out = Vec::<Point>::new();
        let partitions = self.partitions_between(from, to, order);
        let mut shards = HashSet::<u32>::new();
        for tag in tag_ids {
            shards.insert(shard_for_tag(*tag, self.cfg.storage.shard_count));
        }

        for p in partitions {
            for shard in &shards {
                let seg_path = crate::storage::layout::segment_path(&self.cfg.storage.data_dir, &p, *shard);
                let idx_path = crate::storage::layout::index_path(&self.cfg.storage.data_dir, &p, *shard);
                if !seg_path.exists() {
                    continue;
                }
                let ranges = self.scan_ranges(&idx_path, &seg_path, tag_ids, from, to).await?;
                let ranges: Vec<(u64, u64)> = if order == QueryOrder::Desc {
                    ranges.into_iter().rev().collect()
                } else {
                    ranges
                };
                for (start, end) in ranges {
                    if out.len() >= limit {
                        return Ok(RawResult { points: out, truncated: true });
                    }
                    let data = read_range(&seg_path, start, end).await?;
                    let mut recs = decode_points(&data, &tag_set, from, to);
                    if order == QueryOrder::Desc {
                        recs.reverse();
                    }
                    for r in recs {
                        if out.len() >= limit {
                            return Ok(RawResult { points: out, truncated: true });
                        }
                        out.push(r);
                    }
                }
            }
        }

        out.sort_by(|a, b| {
            if a.ts_epoch == b.ts_epoch {
                a.tag_id.cmp(&b.tag_id)
            } else if order == QueryOrder::Asc {
                a.ts_epoch.cmp(&b.ts_epoch)
            } else {
                b.ts_epoch.cmp(&a.ts_epoch)
            }
        });
        Ok(RawResult { points: out, truncated: false })
    }

    pub async fn range(
        &self,
        tag_ids: &[u32],
        from: i64,
        to: i64,
        bucket_ms: Option<i64>,
        agg: AggName,
        order: QueryOrder,
        limit: usize,
    ) -> Result<RangeResult> {
        if bucket_ms.is_none() {
            let raw = self.raw(tag_ids, from, to, limit, order).await?;
            let mut buckets = HashMap::new();
            for p in raw.points {
                buckets
                    .entry(p.tag_id.to_string())
                    .or_insert_with(Vec::new)
                    .push(BucketValue {
                        bucket_start: p.ts_epoch.to_string(),
                        value: point_to_json(&p.value),
                    });
            }
            return Ok(RangeResult {
                buckets,
                truncated: raw.truncated,
            });
        }
        let raw = self.raw(tag_ids, from, to, limit, order).await?;
        let bucket_ms = bucket_ms.unwrap();
        let step = match self.cfg.storage.timestamp_unit {
            TimestampUnit::Us => bucket_ms * 1_000,
            TimestampUnit::Ns => bucket_ms * 1_000_000,
        };
        let mut states: HashMap<u32, BTreeMap<i64, AggState>> = HashMap::new();
        for p in raw.points {
            let b = ((p.ts_epoch - from) / step) * step + from;
            let entry = states
                .entry(p.tag_id)
                .or_default()
                .entry(b)
                .or_insert_with(AggState::default);
            entry.update(&p);
        }
        let mut out = HashMap::<String, Vec<BucketValue>>::new();
        for (tag, bmap) in states {
            let mut rows = Vec::new();
            for (bucket_start, st) in bmap {
                rows.push(BucketValue {
                    bucket_start: bucket_start.to_string(),
                    value: st.finalize(agg),
                });
            }
            if order == QueryOrder::Desc {
                rows.reverse();
            }
            out.insert(tag.to_string(), rows);
        }
        Ok(RangeResult {
            buckets: out,
            truncated: raw.truncated,
        })
    }

    fn partitions_between(&self, from: i64, to: i64, order: QueryOrder) -> Vec<PartitionInfo> {
        let from_ms = ts_to_ms(from, &self.cfg.storage.timestamp_unit);
        let to_ms = ts_to_ms(to, &self.cfg.storage.timestamp_unit);
        let step = self.cfg.storage.partition_duration_ms;
        let start = (from_ms / step) * step;
        let mut out = Vec::new();
        let mut cur = start;
        while cur <= to_ms {
            let dt = chrono::DateTime::<chrono::Utc>::from_timestamp_millis(cur).unwrap_or_else(chrono::Utc::now);
            out.push(PartitionInfo {
                start_ms: cur,
                day: dt.format("%Y-%m-%d").to_string(),
                hour: dt.format("%H").to_string(),
            });
            cur += step;
        }
        if order == QueryOrder::Desc {
            out.reverse();
        }
        out
    }

    async fn scan_ranges(
        &self,
        idx_path: &Path,
        seg_path: &Path,
        tag_ids: &[u32],
        from: i64,
        to: i64,
    ) -> Result<Vec<(u64, u64)>> {
        let min_tag = *tag_ids.iter().min().unwrap_or(&0);
        let max_tag = *tag_ids.iter().max().unwrap_or(&u32::MAX);
        if idx_path.exists() {
            let idx = tokio::fs::read(idx_path).await?;
            let entries = decode_block_index(&idx);
            let mut ranges = Vec::new();
            for e in entries {
                if e.max_ts < from || e.min_ts > to {
                    continue;
                }
                if e.max_tag_id < min_tag || e.min_tag_id > max_tag {
                    continue;
                }
                ranges.push((e.byte_offset_start, e.byte_offset_end));
            }
            return Ok(ranges);
        }
        let sz = tokio::fs::metadata(seg_path).await?.len();
        Ok(if sz > 0 { vec![(0, sz)] } else { vec![] })
    }
}

async fn read_range(path: &Path, start: u64, end: u64) -> Result<Vec<u8>> {
    if end <= start {
        return Ok(Vec::new());
    }
    let data = tokio::fs::read(path).await?;
    let s = start as usize;
    let e = end as usize;
    Ok(data[s..e.min(data.len())].to_vec())
}

fn decode_points(buf: &[u8], tags: &HashSet<u32>, from: i64, to: i64) -> Vec<Point> {
    let mut out = Vec::new();
    let mut off = 0usize;
    while off < buf.len() {
        if let Some((p, n)) = decode_segment_record(buf, off) {
            off += n;
            if !tags.contains(&p.tag_id) {
                continue;
            }
            if p.ts_epoch < from || p.ts_epoch > to {
                continue;
            }
            out.push(p);
        } else {
            break;
        }
    }
    out
}

fn point_to_json(v: &PointValue) -> serde_json::Value {
    match v {
        PointValue::Num(n) => serde_json::json!(n),
        PointValue::Str(s) => serde_json::json!(s),
    }
}

#[derive(Default, Clone)]
struct AggState {
    first: Option<PointValue>,
    last: Option<PointValue>,
    min: Option<f64>,
    max: Option<f64>,
    sum: f64,
    numeric_count: u64,
    count: u64,
}

impl AggState {
    fn update(&mut self, p: &Point) {
        self.count += 1;
        if self.first.is_none() {
            self.first = Some(p.value.clone());
        }
        self.last = Some(p.value.clone());
        if p.type_code.is_numeric() {
            if let PointValue::Num(v) = &p.value {
                self.min = Some(self.min.map(|x| x.min(*v)).unwrap_or(*v));
                self.max = Some(self.max.map(|x| x.max(*v)).unwrap_or(*v));
                self.sum += *v;
                self.numeric_count += 1;
            }
        }
    }
    fn finalize(&self, agg: AggName) -> serde_json::Value {
        match agg {
            AggName::Count => serde_json::json!(self.count),
            AggName::First => match &self.first {
                Some(PointValue::Num(v)) => serde_json::json!(v),
                Some(PointValue::Str(v)) => serde_json::json!(v),
                None => serde_json::Value::Null,
            },
            AggName::Last => match &self.last {
                Some(PointValue::Num(v)) => serde_json::json!(v),
                Some(PointValue::Str(v)) => serde_json::json!(v),
                None => serde_json::Value::Null,
            },
            AggName::Min => self.min.map(serde_json::Value::from).unwrap_or(serde_json::Value::Null),
            AggName::Max => self.max.map(serde_json::Value::from).unwrap_or(serde_json::Value::Null),
            AggName::Avg => {
                if self.numeric_count == 0 {
                    serde_json::Value::Null
                } else {
                    serde_json::json!(self.sum / self.numeric_count as f64)
                }
            }
        }
    }
}
